include/ada/url_base.h

```diff
@@ -73,55 +73,55 @@ namespace ada {
      * Inspired after servo/url
      * @see https://github.com/servo/rust-url/blob/b65a45515c10713f6d212e6726719a020203cc98/url/src/quirks.rs#L31
      */
-    [[nodiscard]] ada_really_inline ada::url_components get_components() noexcept;
+    [[nodiscard]] virtual ada_really_inline ada::url_components get_components() noexcept = 0;
 
     /**
      * The origin getter steps are to return the serialization of this’s URL’s origin. [HTML]
      * @see https://url.spec.whatwg.org/#concept-url-origin
      */
-    [[nodiscard]] std::string get_origin() const noexcept;
+    [[nodiscard]] virtual std::string get_origin() const noexcept = 0;
```

Comment:
It is somewhat unfortunate (though maybe difficult to avoid) that all these get_origin()... return std::string. I think it is fine and I don't suggest you change it, but we have to be mindful that competing libraries return the lightweight std::string_view and they have thus a performance advantage builtin that we don't have.